README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# predmark

<!-- badges: start -->

[![Lifecycle:
experimental](https://img.shields.io/badge/lifecycle-experimental-orange.svg)](https://www.tidyverse.org/lifecycle/#maturing)
[![CRAN
status](https://www.r-pkg.org/badges/version/predmark)](https://CRAN.R-project.org/package=predmark)
[![Codecov test
coverage](https://codecov.io/gh/kiernann/predmark/branch/master/graph/badge.svg)](https://codecov.io/gh/kiernann/predmark?branch=master)
[![R build
status](https://github.com/kiernann/predmark/workflows/R-CMD-check/badge.svg)](https://github.com/kiernann/predmark/actions)
<!-- badges: end -->

This package contains data and documentation from prediction markets and
forecasting models for the 2018 and 2020 U.S. elections. The data is
formatted in ways to facilitate statistical comparison. This package was
built to both hold the data and make for easy installation, including
package dependencies needed to completely replicate the data formatting.

The [original repository](https://github.com/kiernann/models-markets)
for this project was not formatted as a package.

## Installation

You can install the development version of predmark from
[GitHub](https://github.com/kiernann/predmark) with:

``` r
# install.packages("remotes")
remotes::install_github("kiernann/predmark")
```

## Data

``` r
library(tidyverse)
library(predmark)
```

The package contains daily predictions from PredictIt and
FiveThirtyEight.

``` r
markets18
#> # A tibble: 41,933 x 9
#>    date       mid   race  party  open   low  high close volume
#>    <date>     <chr> <chr> <chr> <dbl> <dbl> <dbl> <dbl>  <dbl>
#>  1 2017-01-27 2918  MA-S1 D      0.79  0.71  0.86  0.74   1102
#>  2 2017-01-28 2918  MA-S1 D      0.74  0.74  0.78  0.78   1010
#>  3 2017-01-29 2918  MA-S1 D      0.78  0.76  0.78  0.77    581
#>  4 2017-01-30 2918  MA-S1 D      0.77  0.76  0.78  0.78    631
#>  5 2017-01-31 2918  MA-S1 D      0.78  0.77  0.81  0.81   1378
#>  6 2017-02-01 2918  MA-S1 D      0.81  0.79  0.82  0.8     768
#>  7 2017-02-02 2918  MA-S1 D      0.8   0.79  0.8   0.79     50
#>  8 2017-02-03 2918  MA-S1 D      0.79  0.78  0.8   0.78    592
#>  9 2017-02-04 2918  MA-S1 D      0.78  0.78  0.79  0.79     10
#> 10 2017-02-05 2918  MA-S1 D      0.79  0.79  0.8   0.8       6
#> # … with 41,923 more rows
```

``` r
model18
#> # A tibble: 1,049 x 11
#>    date       race  name       party chamber special incumbent   prob voteshare min_share max_share
#>    <date>     <chr> <chr>      <chr> <chr>   <lgl>   <lgl>      <dbl>     <dbl>     <dbl>     <dbl>
#>  1 2018-09-25 VA-04 A. Donald… D     house   FALSE   TRUE      0.999     0.632     0.579     0.684 
#>  2 2018-09-25 GA-03 A. Drew F… R     house   FALSE   TRUE      1.00      0.676     0.627     0.725 
#>  3 2018-09-25 LA-03 Aaron And… L     house   FALSE   FALSE     0.0002    0.0254    0.0071    0.0486
#>  4 2018-09-25 ID-02 Aaron Swi… D     house   FALSE   FALSE     0.002     0.364     0.309     0.418 
#>  5 2018-09-25 IA-01 Abby Fink… D     house   FALSE   FALSE     0.971     0.554     0.506     0.600 
#>  6 2018-09-25 VA-07 Abigail S… D     house   FALSE   FALSE     0.352     0.474     0.431     0.517 
#>  7 2018-09-25 IL-16 Adam Kinz… R     house   FALSE   TRUE      0.984     0.590     0.538     0.642 
#>  8 2018-09-25 CA-28 Adam Schi… D     house   FALSE   TRUE      1         0.812     0.766     0.857 
#>  9 2018-09-25 WA-09 Adam Smith D     house   FALSE   TRUE      0.982     0.613     0.544     0.683 
#> 10 2018-09-25 NE-03 Adrian Sm… R     house   FALSE   TRUE      1         0.716     0.664     0.768 
#> # … with 1,039 more rows
```

<!-- refs: start -->

<!-- refs: end -->


NEWS.md
# predmark 1.0.0

* Added a `NEWS.md` file to track changes to the package.
* Copy the 2018 code and data from `models-markets`.


R/memento.R
#' Write a memento file
#'
#' Read an archived file and write it to a local directory.
#'
#' @param url URL to retrieve information for.
#' @param date The date of the archive.
#' @param dir The directory in which to save the file.
#' @importFrom here here
#' @importFrom fs file_exists dir_create as_fs_path
#' @importFrom readr write_lines
#' @return
#' The path of the created file, invisibly.
#' @export
write_memento <- function(url, date, dir) {
  path <- fs::path(dir, basename(url))
  if (!fs::file_exists(path)) {
    data <- read_memento(url, date, as = "text", encoding = "UTF-8")
    fs::dir_create(dirname(path))
    readr::write_lines(data, path)
  }
  invisible(fs::as_fs_path(path))
}

#' Read a file from the Time Travel API
#'
#' This function is adapted from the hrbrmstr/wayback package on GitHub.
#'
#' @param url URL to retrieve information for
#' @param date Date to use when checking for availability. If you don't
#'   pass in a valid R "time-y" object, you will need to ensure the character
#'   string you provide is in a valid subset of `YYYYMMDDhhmmss`.
#' @param as How you want the content returned. One of "`text`", "`raw`" or
#'   "`parsed`" (it uses [httr::content()] to do the heavy lifting).
#' @param ... Additional arguments passed to [httr::content()].
#' @source <https://github.com/hrbrmstr/wayback/blob/master/R/mementoweb.r>
#' @importFrom httr GET stop_for_status content
#' @importFrom glue glue
#' @return The content from the GET request.
#' @export
read_memento <- function(url, date = format(Sys.Date(), "%Y"),
                         as = c("text", "raw", "parsed"), ...) {
  as <- match.arg(as, c("text", "raw", "parsed"))
  if (is.null(timestamp)) {
    timestamp <- format(Sys.Date(), "%Y")
  }
  if (is.character(timestamp) && (timestamp == "")) {
    timestamp <- format(Sys.Date(), "%Y")
  }
  if (inherits(timestamp, "POSIXct")) {
    timestamp <- format(timestamp, "%Y%m%d%H%M%S")
  }
  url <- glue::glue("http://timetravel.mementoweb.org/memento/{date}/{span}")
  res <- httr::GET(url)
  httr::stop_for_status(res)
  httr::content(res, as = as)
}


data-raw/model18.R
## code to prepare `model18` dataset goes here

library(stringr)
library(readr)
library(dplyr)
library(tidyr)
library(here)

# read model and results data from https://fivethirtyeight.com ------------

## District level 538 House model history
## Updated:  2018-11-06 at 01:56
## Archived: 2018-11-06 at 12:06
write_memento(
  url = "https://projects.fivethirtyeight.com/congress-model-2018/house_district_forecast.csv",
  date = "2018-11-06",
  dir = here("data-raw", "models")
)

house_district_forecast <- read_csv(
  file = here("data-raw", "models", "house_district_forecast.csv"),
  col_types = cols(
    forecastdate = col_date(),
    state = col_character(),
    district = col_double(),
    special = col_logical(),
    candidate = col_character(),
    party = col_character(),
    incumbent = col_logical(),
    model = col_character(),
    win_probability = col_double(),
    voteshare = col_double(),
    p10_voteshare = col_double(),
    p90_voteshare = col_double()
  )
)

## Seat level 538 Senate model history
## Updated:  2018-11-06 at 11:06
## Archived: 2018-11-06 at 21:00
write_memento(
  url = "https://projects.fivethirtyeight.com/congress-model-2018/senate_seat_forecast.csv",
  date = "2018-11-06",
  dir = here("data-raw", "models")
)

senate_seat_forecast <- read_csv(
  file = here("data-raw", "models", "senate_seat_forecast.csv"),
  col_types = cols(
    forecastdate = col_date(),
    state = col_character(),
    class = col_double(),
    special = col_logical(),
    candidate = col_character(),
    party = col_character(),
    incumbent = col_logical(),
    model = col_character(),
    win_probability = col_double(),
    voteshare = col_double(),
    p10_voteshare = col_double(),
    p90_voteshare = col_double()
  )
)

# format model history ----------------------------------------------------

## source:    https://fivethirtyeight.com/
## input:     input/*_forecast.csv
## desc:      history of forecasting model top line probabilities
## use:       operationalize probabilistic forecasts from a forecasting model

# format district for race variable
model_district <- house_district_forecast %>%
  mutate(
    district = str_pad(
      string = district,
      width = 2,
      side = "left",
      pad = "0"
    )
  )

# format class for race variable
model_seat <- senate_seat_forecast %>%
  rename(district = class) %>%
  mutate(
    district = str_pad(
      string = district,
      width = 2,
      side = "left",
      pad = "S"
    )
  )

model_combined <-
  bind_rows(model_district, model_seat, .id = "chamber") %>%
  # create race variable for relational join
  unite(
    col = race,
    state, district,
    sep = "-",
    remove = TRUE
  ) %>%
  rename(
    name = candidate,
    date = forecastdate,
    prob = win_probability,
    min_share = p10_voteshare,
    max_share = p90_voteshare
  ) %>%
  mutate(
    chamber = recode(
      .x = chamber,
      "1" = "house",
      "2" = "senate"
    ),
    # only special elections are for senate.
    special = case_when(
      is.na(special) ~ FALSE,
      !is.na(special) ~ special
    ),
    # both caucus with Democrats
    party = case_when(
      name == "Bernard Sanders" ~ "D",
      name == "Angus S. King Jr." ~ "D",
      party == "LIB" ~ "L",
      party == "IND" ~ "I",
      party == "GRE" ~ "G",
      TRUE ~ party
    )
  ) %>%
  mutate(across(prob, round, digits = 4)) %>%
  # convert percent vote share values to decimal
  mutate(across(10:12, ~round(. * 0.01, digits = 4))) %>%
  # keep only named candidates
  filter(name != "Others", name != "Zak Ringelstein") %>%
  # reorder data frame
  select(date, race, name, party, chamber, everything()) %>%
  arrange(date, name)

# separate model data by model format
model_combined <- model_combined %>%
  group_split(model) %>%
  set_names(c("classic", "lite", "deluxe")) %>%
  map(select, -model)

# according to 538, the "classic" model can be used as a default
model18 <- model_combined$classic

# export model data -------------------------------------------------------

write_tsv(model18, here("data-raw", "model18.tsv"))
usethis::use_data(model18, overwrite = TRUE)


data-raw/markets18.R
## code to prepare `markets18` dataset goes here

library(stringr)
library(readr)
library(dplyr)
library(tidyr)
library(here)

# read market data from https://www.predictit.org/ ------------------------

## Market Data sent by will.jennings@predictit.org
## Detailed price history provided to academic researchers
DailyMarketData <- read_delim(
  file = here("data-raw", "markets", "DailyMarketData.csv"),
  delim = "|",
  na = "n/a",
  col_types = cols(
    MarketId = col_character(),
    ContractName = col_character(),
    ContractSymbol = col_character(),
    Date = col_date(format = "")
  )
)

Market_ME02 <- read_csv(
  file = here("data-raw", "markets", "Market_ME02.csv"),
  col_types = cols(
    ContractID = col_character(),
    Date = col_date(format = "%m/%d/%Y")
  )
)

Contract_NY27 <- read_csv(
  file = here("data-raw", "markets", "Contract_NY27.csv"),
  na = c("n/a", "NA"),
  skip = 156, # this file was a mess
  col_types = cols(
    ContractID = col_character(),
    Date = col_date(format = "%m/%d/%Y")
  )
)

# format markets history ------------------------------------------------------

## source:    https://predictit.org/
## input:     data/raw/models/DailyMarketData.csv
## desc:      history of contract prices for midterm election markets
## use:       operationalize probabalistic forecasts from prediction markets

markets18 <- DailyMarketData %>%
  rename(
    mid    = MarketId,
    name   = MarketName,
    symbol = MarketSymbol,
    party  = ContractName,
    open   = OpenPrice,
    close  = ClosePrice,
    high   = HighPrice,
    low    = LowPrice,
    volume = Volume,
    date   = Date
  ) %>%
  select(date, everything()) %>%
  select(-ContractSymbol)

# get candidate names from full market question
markets18$name[str_which(markets18$name, "Which party will")] <- NA
markets18$name <- word(markets18$name, start = 2, end = 3)

# recode party variables
markets18$party <- recode(
  .x = markets18$party,
  "Democratic or DFL" = "D",
  "Democratic" = "D",
  "Republican" = "R"
)

# remove year information from symbol strings
markets18 <- markets18 %>%
  mutate(
    race = symbol %>%
      str_remove("\\.\\d{2,4}$") %>% # year
      str_remove("(.*)\\.") %>% # names
      str_replace("(?<=[:upper:])SE(.*)$", "S1") %>% # senate
      str_replace("AL", "01") %>% # at large
      str_remove("((?<=\\d))G") %>% # general?
      str_replace("99$", "S2"), # special election
    race = case_when(
      name == "SPEC" ~ "MSS2",
      mid == "3857" ~ "CAS1",
      TRUE ~ race
    ),
    name = case_when(
      name == "PARTY" ~ NA_character_,
      name == "SPEC" ~ NA_character_,
      TRUE ~ name
    ),
    race = race %>%
      str_replace(
        pattern = "([:alpha:]{2})(S\\d|\\d{2})",
        replacement = "\\1-\\2"
      )
  ) %>%
  select(-symbol) %>%
  # remove markets incorectly repeated
  # some not running for re-election
  filter(
    mid != "3455", # Paul Ryan
    mid != "3507", # Jeff Flake
    mid != "3539", # Shea-Porter
    mid != "3521", # Darrell Issa
    mid != "3522", # Repeat of 4825
    mid != "4177", # Repeat of 4232
    mid != "4824"  # Repeat of 4776
  )

# divide the data based on market question syntax
# market questions provided name or party, never both
markets_with_name <- markets18 %>%
  filter(is.na(party)) %>%
  select(-party)

markets_with_party <- markets18 %>%
  filter(is.na(name)) %>%
  select(-name)

# join with members key to add party, then back with rest of market
markets18 <- markets_with_name %>%
  inner_join(members115, by = c("name", "race")) %>%
  select(date, mid, race, party, open, low, high, close, volume) %>%
  bind_rows(markets_with_party)

# add in ME-02 and NY-27 which were left out of initial data
ny_27 <- Contract_NY27 %>%
  rename_all(str_to_lower) %>%
  slice(6:154) %>%
  mutate(
    mid = "4729",
    race = "NY-27",
    party = "R"
  ) %>%
  select(-average)

me_02 <- Market_ME02 %>%
  rename_all(str_to_lower) %>%
  rename(party = longname) %>%
  filter(date != "2018-10-10") %>%
  mutate(mid = "4945", race = "ME-02")

markets_extra <-
  bind_rows(ny_27, me_02) %>%
  select(date, mid, race, party, open, low, high, close, volume)

markets_extra$party[str_which(markets_extra$party, "GOP")] <- "R"
markets_extra$party[str_which(markets_extra$party, "Dem")] <- "D"

# bind with ME-02 and NY-27
markets18 <- bind_rows(markets18, markets_extra)

# export market data ------------------------------------------------------

write_tsv(markets18, here("data-raw", "markets18.tsv"))
usethis::use_data(markets18, overwrite = TRUE)


data-raw/results18.R
## code to prepare `results18` dataset goes here

# save archived results ---------------------------------------------------

# Midterm election results via ABC and 538
# Used in https://53eig.ht/2PiFb0f
# Published: 2018-12-04 at 17:56
# Archived:  2018-04-04 at 16:08

# Midterm election results via ABC and 538
# Used in https://53eig.ht/2PiFb0f
# Published: 2018-12-04 at 17:56
# Archived:  2018-04-04 at 16:08
write_memento(
  url = paste(
    "https://raw.githubusercontent.com/fivethirtyeight/data/master",
    "forecast-review",
    "forecast_results_2018.csv",
    sep = "/"
  ),
  date = "2019-04-04",
  dir = here("data-raw", "results")
)

forecast_results <- read_csv(
  file = here::here("data-raw", "results", "forecast_results_2018.csv"),
  col_types  = cols(
    cycle = col_integer(),
    Democrat_Won = col_logical(),
    Republican_Won = col_logical(),
    uncalled = col_logical(),
    forecastdate = col_date(format = "%m/%d/%y"),
    category = col_factor(
      ordered = TRUE,
      levels = c(
        "Solid D",
        "Likely D",
        "Lean D",
        "Tossup (Tilt D)",
        "Tossup (Tilt R)",
        "Lean R",
        "Likely R",
        "Safe R"
      )
    )
  )
)

# format election results -------------------------------------------------

## source:    https://fivethirtyeight.com/
## input:     input/forecast_results_2018.csv
## desc:      final predictions and election results
## use:       assess the accuracy of both predictive methods

results18 <- forecast_results %>%
  filter(
    branch != "Governor",
    version == "classic",
    uncalled == FALSE
  ) %>%
  separate(
    col = race,
    into = c("state", "district"),
    sep = "-"
  ) %>%
  rename(winner = Democrat_Won) %>%
  mutate(
    district = str_pad(district, width = 2,  pad   = "0"),
    category = fct_recode(
      .f = category,
      "Tilt D" = "Tossup (Tilt D)",
      "Tilt R" = "Tossup (Tilt R)"
      "Like D" = "Likely D"
      "Like R" = "Likely R"
    )
  ) %>%
  unite(
    state, district,
    col = race,
    sep = "-"
  ) %>%
  select(year = cycle, race, category, winner)


# add partisan lean -------------------------------------------------------

# Average difference between how a district votes and the country
# Updated:  2018-11-19 at 16:13
# Archived: 2018-04-04 at 16:05
write_memento(
  date = "2019-04-04",
  dir = here("data-raw", "results"),
  url = paste(
    "https://raw.githubusercontent.com/fivethirtyeight/data/master",
    "partisan-lean",
    "fivethirtyeight_partisan_lean_DISTRICTS.csv",
    sep = "/"
  )
)

partisan_lean_DISTRICTS <- read_csv(
  file = here("data-raw", "results", "fivethirtyeight_partisan_lean_DISTRICTS.csv"),
  col_types = cols(
    district = col_character(),
    pvi_538 = col_character()
  )
)

write_memento(
  date = "2019-04-04",
  dir = here("data-raw", "results"),
  url = paste(
    "https://raw.githubusercontent.com/fivethirtyeight/data/master",
    "partisan-lean",
    "fivethirtyeight_partisan_lean_STATES.csv",
    sep = "/"
  )
)

partisan_lean_STATES <- read_csv(
  file = here("data-raw", "results", "fivethirtyeight_partisan_lean_STATES.csv"),
  col_types = cols(
    state = col_character(),
    pvi_538 = col_character()
  )
)

# Separate lean value and replace state name with state abbreviation
lean_states <- partisan_lean_STATES %>%
  separate(
    col = pvi_538,
    into = c("party", "lean"),
    sep = "\\+",
    convert = TRUE
  ) %>%
  rename(race = state) %>%
  mutate(race = paste(state.abb, "S1", sep = "-"))

# Seperate lean value and pad district number for race code
lean_district <- partisan_lean_DISTRICTS %>%
  separate(
    col = pvi_538,
    into = c("party", "lean"),
    sep = "\\+",
    convert = TRUE
  ) %>%
  separate(
    col = district,
    into = c("state", "race"),
    sep = "\\-"
  ) %>%
  mutate(race = str_pad(race, width = 2, pad = "0")) %>%
  unite(
    col = race,
    state, race,
    sep = "-"
  )

# Turn single number into negative-positive spectrum
lean18 <-
  bind_rows(lean_states, lean_district) %>%
  mutate(
    lean = case_when(
      party == "D" ~ lean * -1,
      party == "R" ~ lean
    )
  )

results18 <- left_join(
  x = results18,
  y = lean18,
  by = "race"
)

results18 <- relocate(results18, lean, .before = category)

# export results ----------------------------------------------------------

write_tsv(results18, here("data-raw", "results18.tsv"))
usethis::use_data(results18, overwrite = TRUE)


data-raw/members115.R
## code to prepare `members18` dataset goes here

library(stringr)
library(readr)
library(dplyr)
library(tidyr)
library(here)

# save the archived raw data ----------------------------------------------

## Current members of the 115th
## Archived: 2018-10-22 at 18:11
write_memento(
  url = "https://theunitedstates.io/congress-legislators/legislators-current.csv",
  date = "2018-10-22",
  dir = here("data-raw", "members")
)

legislators_current <- read_csv(
  file = here("data-raw", "members", "legislators-current.csv"),
  col_types = cols(
    birthday = col_date(),
    govtrack_id = col_character()
  )
)

# The ideology and leadership scores of the 115th
# Calculated with cosponsorship analysis
# Archived 2019-01-21 17:13:08
write_memento(
  url = "https://www.govtrack.us/data/analysis/by-congress/115/sponsorshipanalysis_h.txt",
  date = "2019-03-23",
  dir = here("data-raw", "members")
)

sponsorshipanalysis_h <- read_csv(
  file = here("data-raw", "members", "sponsorshipanalysis_h.txt"),
  col_types = cols(
    ID = col_character()
  )
)

write_memento(
  url = "https://www.govtrack.us/data/analysis/by-congress/115/sponsorshipanalysis_s.txt",
  date = "2019-03-23",
  dir = here("data-raw", "members")
)

sponsorshipanalysis_s <- read_csv(
  file = here("data-raw", "members", "sponsorshipanalysis_s.txt"),
  col_types = cols(
    ID = col_character()
  )
)

# format member data ------------------------------------------------------

## source:    https://theunitedstates.io/
## input:     data/raw/members/legislators_current.csv
## desc:      members of the 115th Congress w/ bio and pol info
## use:       supplement prediction history and contextualize election results

members115 <- legislators_current %>%
  unite(
    # create single name
    first_name, last_name,
    col = name,
    sep = " "
  ) %>%
  rename(
    gid     = govtrack_id,
    chamber = type,
    class   = senate_class,
    birth   = birthday
  ) %>%
  select(
    name,
    gid,
    birth,
    state,
    district,
    class,
    party,
    gender,
    chamber
  ) %>%
  arrange(
    chamber
  )

# recode, encode, and pad
members115 <- members115 %>%
  mutate(
    # fix names to match 538
    name = name %>%
      iconv(to = "ASCII//TRANSLIT") %>%
      str_replace_all("Robert Menendez", "Bob Menendez") %>%
      str_replace_all("Robert Casey",    "Bob Casey") %>%
      str_replace_all("Bernard Sanders", "Bernie Sanders"),
    chamber = recode(chamber, "rep" = "house", "sen" = "senate"),
    district = str_pad(district, width = 2, pad = "0"),
    class = str_pad(class, width = 2, pad = "S"),
    party = recode(
      .x = party,
      "Democrat" = "D",
      "Independent" = "D",
      "Republican" = "R"
    ),
    district = if_else(
      condition = is.na(district),
      true = class,
      false = district
    )
  ) %>%
  # create district code as relational key
  unite(
    col = race,
    state, district,
    sep = "-",
    remove = TRUE
  ) %>%
  select(-class) %>%
  arrange(name)

# format member stats for join
members_stats <-
  bind_rows(
    sponsorshipanalysis_h,
    sponsorshipanalysis_s,
    .id = "chamber"
  ) %>%
  select(
    ID,
    chamber,
    party,
    ideology,
    leadership
  ) %>%
  rename(gid = ID) %>%
  mutate(
    gid = as.character(gid),
    chamber = recode(chamber, "1" = "house", "2" = "senate"),
    party = recode(
      .x = party,
      "Democrat" = "D",
      "Independent" = "D",
      "Republican" = "R"
    )
  )

# add stats to frame by GovTrack ID
members115 <- inner_join(
  x = members115,
  y = members_stats,
  by = c("gid", "party", "chamber")
)

# export member data ------------------------------------------------------

write_tsv(members115, here("data-raw", "members115.tsv"))
usethis::use_data(members115, overwrite = TRUE)


R/data.R
#' Members of the 115th Congress
#'
#' The biographical information of the members is taken from an archived
#' version of the `current-legislators` data from the UnitedStates.io project.
#' The ideology and leadership scores are taken from GovTrack, which bases their
#' analysis on the bills introduced and cosponsored by members.
#'
#' @format A data frame with 534 rows and 9 variables:
#' \describe{
#'   \item{name}{Legislator full name}
#'   \item{gid}{Unique GovTrack ID number}
#'   \item{birth}{Birth date}
#'   \item{race}{District or senate state and class code}
#'   \item{party}{Political party}
#'   \item{gender}{Identifying gender}
#'   \item{chamber}{Chamber of Congress: House or Senate}
#'   \item{ideology}{Left-Right (0-1) ideology score from pattern of cosponsorship}
#'   \item{leadership}{Leadership score based on frequency of cosponsored bills}
#'   ...
#' }
#' @source <https://theunitedstates.io/congress-legislators>
"members115"

#' Daily trading prices of 112 competitive 2018 Midterm races
#'
#' The opening, low, high, and closing prices of contracts tied to a party
#' winning a midterm race. Equilibrium prices reflect outcome probability.
#'
#' Prediction markets generate probabilistic forecasts by crowd-sourcing the
#' collection of data from self-interested and risk averse traders. The
#' efficient market hypothesis holds that asset prices reflect all available
#' information (including forecasting models).
#'
#' PredictIt is an exchange run by Victoria University of Wellington, New
#' Zealand. The site offers a continuous double-auction exchange, where traders
#' buy and sell shares of futures contracts tied to election outcomes. As a
#' trader’s perception of probabilities changes, they can sell those shares. The
#' market equilibrium price updates accordingly to reflect current probability.
#' As outcomes become more likely, prices rise as demand for shares increases.
#' Market data for all midterm prices was provided for academic purposes.
#'
#' @format A data frame with 41,933 rows and 9 variables:
#' \describe{
#'   \item{date}{Prediction date}
#'   \item{mid}{Market ID}
#'   \item{race}{Unique race code}
#'   \item{party}{Contract party}
#'   \item{open}{Opening contract price for day}
#'   \item{low}{Low contract price for day}
#'   \item{high}{High contract price for day}
#'   \item{close}{Closing contract price for day}
#'   \item{volume}{Nunber of contract shares traded}
#'   ...
#' }
#' @source <https://www.predictit.org/>
"markets18"

#' Daily forecast probabilities of all 2018 Midterm races
#'
#' The election forecast probabilities from the 2018 FiveThirtyEight "classic"
#' midterm model. Probabilities are taken by estimating vote share and a
#' probability distribution from historical data.
#'
#' As Nate Silver explains, most forecasting models (1) “take lots of polls,
#' perform various types of adjustments to them, and then blend them with other
#' kinds of empirically useful indicators to forecast each race”. Importantly,
#' they (2) “account for the uncertainty in the forecast and simulate the
#' election thousands of times” to generate a probabilistic forecast.
#'
#' The classic model incorporates three types of inputs, primarily direct and
#' imputed polling as well as fundamentals factors like incumbency and the
#' generic ballot.
#'
#' @format A data frame with 1,049 rows and 11 variables:
#' \describe{
#'   \item{date}{Prediction date}
#'   \item{race}{Unique race code}
#'   \item{name}{Candidate name}
#'   \item{party}{Candidate party}
#'   \item{chamber}{Chamber of Congress: House or Senate}
#'   \item{special}{Is this election a special election?}
#'   \item{incumbent}{Is the candidate the incumbent?}
#'   \item{prob}{Probability candidate wins}
#'   \item{voteshare}{Median estimated share of vote}
#'   \item{min_share}{10th percentile vote share}
#'   \item{max_share}{90th percentile vote share}
#'   ...
#' }
#' @source <https://fivethirtyeight.com/>
"model18"

#' 2018 Midterm race results
#'
#' @format A data frame with 468 rows and 6 variables:
#' \describe{
#'   \item{year}{Election year}
#'   \item{race}{Unique race code}
#'   \item{lean}{The average difference between how a state or district votes
#'   and how the country votes overall}
#'   \item{category}{Ordinal race category: safe > solid > likely > lean > tilt}
#'   \item{winner}{`TRUE` if Democrat won, `FALSE` if Republican}
#'   \item{party}{Incumbent party}
#'   ...
#' }
#' @source <https://fivethirtyeight.com/>
"results18"
